File: ai_player.py
def minimax(board, depth, is_maximizing):
    scores = {"X": -1, "O": 1, "Tie": 0}
    
    winner = board.check_winner()
    if winner:
        return scores[winner]
    if board.is_full():
        return scores["Tie"]
        
    if is_maximizing:
        best_score = float('-inf')
        for pos in board.get_empty_positions():
            board.cells[pos] = "O"
            score = minimax(board, depth + 1, False)
            board.cells[pos] = " "
            best_score = max(score, best_score)
        return best_score
    else:
        best_score = float('inf')
        for pos in board.get_empty_positions():
            board.cells[pos] = "X"
            score = minimax(board, depth + 1, True)
            board.cells[pos] = " "
            best_score = min(score, best_score)
        return best_score

def get_best_move(board):
    best_score = float('-inf')
    best_move = None
    
    for pos in board.get_empty_positions():
        board.cells[pos] = "O"
        score = minimax(board, 0, False)
        board.cells[pos] = " "
        if score > best_score:
            best_score = score
            best_move = pos
                
    return best_move

File: src/ai-player.js
export function getBestMove(game) {
    const MAX_DEPTH = 6;
    
    function minimax(game, depth, isMaximizing) {
        const winner = game.checkWinner();
        if (winner === "O") return 10 - depth;
        if (winner === "X") return depth - 10;
        if (game.isBoardFull() || depth >= MAX_DEPTH) return 0;

        const scores = [];
        const moves = [];
        const positions = game.getEmptyPositions();

        for (const pos of positions) {
            const gameCopy = game.clone();
            gameCopy.makeMove(pos);
            
            const score = minimax(gameCopy, depth + 1, !isMaximizing);
            scores.push(score);
            moves.push(pos);
        }

        if (moves.length === 0) return 0;

        if (isMaximizing) {
            const maxScore = Math.max(...scores);
            return depth === 0 ? moves[scores.indexOf(maxScore)] : maxScore;
        } else {
            const minScore = Math.min(...scores);
            return depth === 0 ? moves[scores.indexOf(minScore)] : minScore;
        }
    }

    const bestMove = minimax(game.clone(), 0, true);
    return typeof bestMove === 'number' ? bestMove : game.getEmptyPositions()[0];
}